File: apps/burner-windows/src/VM/QemuHost.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;

namespace Flagship.Burner.VM;

/// <summary>
/// The ONE file that runs QEMU. Deliberately dumb: it translates a pure
/// VMConfig (all decisions already made — see QemuCommandLine) into a
/// process and starts/stops it. The pure layer is what's unit-tested; this
/// adapter is exercised by the live e2e boot.
///
/// The analog of the Mac's VZHost: onGuestStopped fires when the guest stops
/// on its own — install completion (the preseed/-no-reboot end the process)
/// or a crash. The pure lifecycle decides what it means (for installs, via
/// the duration-gated VMLifecycle.VerdictForCleanInstallStop).
/// </summary>
public sealed class QemuHost
{
    private readonly QemuToolchain _toolchain;
    private Process? _process;

    public QemuHost(QemuToolchain toolchain)
    {
        _toolchain = toolchain;
    }

    /// <summary>QMP port of the running VM (0 when stopped).</summary>
    public int QmpPort { get; private set; }
    /// <summary>Serial-console port of the running VM (0 when stopped or
    /// production — the console device only exists under a debug grant).</summary>
    public int SerialPort { get; private set; }
    public bool IsRunning => _process is { HasExited: false };

    /// <summary>
    /// Fired (on a thread-pool thread) when the guest stops for ANY reason:
    /// (exitCode, stderrTail). A clean stop is exit 0; the caller marshals to
    /// its own context and feeds the pure lifecycle. Plain settable callback
    /// (not an event) mirroring VZHost.onGuestStopped.
    /// </summary>
    public Action<int, string>? OnGuestStopped { get; set; }

    /// <summary>
    /// Create the sparse main disk (qcow2 grows into it) + the per-VM UEFI
    /// vars copy if they don't exist yet. Idempotent.
    /// </summary>
    public async Task EnsureBundleArtifactsAsync(VMConfig config, VMBundleLayout layout,
                                                 CancellationToken cancellation = default)
    {
        var diskPath = layout.DiskImagePath(config.Name);
        if (!File.Exists(diskPath))
        {
            var gib = config.MainDiskSizeBytes / VMResourcePlan.GiB;
            var psi = new ProcessStartInfo
            {
                FileName = _toolchain.ImgBinary,
                RedirectStandardError = true,
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            psi.ArgumentList.Add("create");
            psi.ArgumentList.Add("-f");
            psi.ArgumentList.Add("qcow2");
            psi.ArgumentList.Add(diskPath);
            psi.ArgumentList.Add($"{gib}G");
            using var p = Process.Start(psi) ?? throw new IOException("qemu-img spawn failed");
            var stderr = await p.StandardError.ReadToEndAsync(cancellation);
            await p.WaitForExitAsync(cancellation);
            if (p.ExitCode != 0)
                throw new IOException($"qemu-img create failed ({p.ExitCode}): {stderr.Trim()}");
        }

        var varsPath = layout.EfiVariableStorePath(config.Name);
        if (!File.Exists(varsPath))
            File.Copy(_toolchain.UefiVarsTemplate, varsPath);
    }

    /// <summary>Start the VM. attachInstallerISO mirrors the lifecycle effects.</summary>
    public async Task StartAsync(VMConfig config, VMBundleLayout layout, bool attachInstallerISO,
                                 string accel = "whpx", CancellationToken cancellation = default)
    {
        if (IsRunning) throw new InvalidOperationException($"VM '{config.Name}' is already running.");
        await EnsureBundleArtifactsAsync(config, layout, cancellation);

        QmpPort = FreeLoopbackPort();
        SerialPort = config.SerialConsoleEnabled ? FreeLoopbackPort() : 0;
        var args = QemuCommandLine.Build(config, layout, _toolchain.UefiCodePath,
                                         attachInstallerISO, QmpPort, SerialPort, accel);

        var psi = new ProcessStartInfo
        {
            FileName = _toolchain.SystemBinary,
            RedirectStandardError = true,
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };
        foreach (var a in args) psi.ArgumentList.Add(a);

        var p = Process.Start(psi) ?? throw new IOException("qemu spawn failed");
        p.EnableRaisingEvents = true;
        var stderrTask = p.StandardError.ReadToEndAsync(CancellationToken.None);
        _ = p.StandardOutput.ReadToEndAsync(CancellationToken.None); // drain, unused
        p.Exited += async (_, _) =>
        {
            string tail = "";
            try { tail = Tail(await stderrTask, 2000); } catch { }
            var code = 0;
            try { code = p.ExitCode; } catch { }
            _process = null;
            QmpPort = 0;
            SerialPort = 0;
            OnGuestStopped?.Invoke(code, tail);
        };
        _process = p;

        // Surface an immediate startup failure (bad args, WHPX refusal) as a
        // thrown error rather than a phantom "running" VM.
        await Task.Delay(750, cancellation);
        if (p.HasExited)
        {
            var stderr = Tail(await stderrTask, 2000);
            throw new IOException($"QEMU failed to start (exit {p.ExitCode}): {stderr.Trim()}");
        }
    }

    /// <summary>
    /// Ask the guest to power down cleanly (ACPI); after graceSeconds without
    /// an exit, hard-stop. The OnGuestStopped event carries the final word.
    /// </summary>
    public async Task StopAsync(int graceSeconds = 90, CancellationToken cancellation = default)
    {
        var p = _process;
        if (p is null || p.HasExited) return;
        try
        {
            await using var qmp = new QmpClient();
            using var connectCts = CancellationTokenSource.CreateLinkedTokenSource(cancellation);
            connectCts.CancelAfter(TimeSpan.FromSeconds(5));
            await qmp.ConnectAsync(QmpPort, connectCts.Token);
            await qmp.SystemPowerdownAsync(connectCts.Token);
        }
        catch
        {
            // QMP unreachable — fall through to the hard stop below.
        }
        var exited = await Task.Run(() => p.WaitForExit(graceSeconds * 1000), cancellation);
        if (!exited) await ForceStopAsync();
    }

    public async Task ForceStopAsync()
    {
        var p = _process;
        if (p is null || p.HasExited) return;
        try
        {
            await using var qmp = new QmpClient();
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(3));
            await qmp.ConnectAsync(QmpPort, cts.Token);
            await qmp.QuitAsync(cts.Token);
            if (await Task.Run(() => p.WaitForExit(3000))) return;
        }
        catch { }
        try { p.Kill(entireProcessTree: true); } catch { }
    }

    private static int FreeLoopbackPort()
    {
        var listener = new TcpListener(IPAddress.Loopback, 0);
        listener.Start();
        var port = ((IPEndPoint)listener.LocalEndpoint).Port;
        listener.Stop();
        return port;
    }

    private static string Tail(string s, int max) => s.Length <= max ? s : s[^max..];
}

File: apps/builder-linux/vm/qemu_host.py
"""The ONE file that runs QEMU.

Deliberately dumb: it translates a pure VMConfig (all decisions already made —
see qemu_command_line.py) into a process and starts/stops it. The pure layer is
what's unit-tested; this adapter is exercised by the live e2e boot. Mirrors
apps/builder-windows/src/VM/QemuHost.cs (the Mac analog is VZHost).

on_guest_stopped fires when the guest stops on its own — install completion
(the preseed/-no-reboot end the process) or a crash. The pure lifecycle decides
what it means (for installs, via the duration-gated verdict).
"""
from __future__ import annotations

import shutil
import socket
import subprocess
import threading
import time
from typing import Callable, Optional

from . import qemu_command_line
from .config import VMConfig
from .inventory import VMBundleLayout
from .qemu_locator import QemuToolchain
from .qmp_client import QmpClient, QmpError
from .resource_plan import GIB


class QemuHostError(Exception):
    pass


def _free_loopback_port() -> int:
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        s.bind(("127.0.0.1", 0))
        return s.getsockname()[1]
    finally:
        s.close()


def _tail(text: str, limit: int = 2000) -> str:
    return text if len(text) <= limit else text[-limit:]


class QemuHost:
    """One live qemu-system-x86_64 process. `accel` is decided by the manager
    ("kvm" normally; "tcg" when the KVM probe said no, with a warning)."""

    def __init__(self, toolchain: QemuToolchain, accel: str = "kvm") -> None:
        self._toolchain = toolchain
        self._accel = accel
        self._proc: Optional[subprocess.Popen] = None
        self.qmp_port = 0
        # 0 when stopped or production — the console device / SSH forward only
        # exist under a debug grant.
        self.serial_port = 0
        self.ssh_port = 0
        # Fired (on a watcher thread) when the guest stops for ANY reason:
        # (exit_code, stderr_tail). The caller marshals to its own context.
        self.on_guest_stopped: Optional[Callable[[int, str], None]] = None

    @property
    def is_running(self) -> bool:
        return self._proc is not None and self._proc.poll() is None

    def ensure_bundle_artifacts(self, config: VMConfig, layout: VMBundleLayout) -> None:
        """Create the sparse main disk (qcow2 grows into it) + the per-VM OVMF
        vars copy if they don't exist yet. Idempotent."""
        import os

        disk = layout.disk_image_path(config.name)
        if not os.path.exists(disk):
            gib = config.main_disk_size_bytes // GIB
            r = subprocess.run(
                [self._toolchain.img_binary, "create", "-f", "qcow2", disk, f"{gib}G"],
                capture_output=True,
                text=True,
            )
            if r.returncode != 0:
                raise QemuHostError(
                    f"qemu-img create failed ({r.returncode}): {r.stderr.strip()}"
                )
        vars_path = layout.efi_variable_store_path(config.name)
        if not os.path.exists(vars_path):
            shutil.copyfile(self._toolchain.uefi_vars_template, vars_path)

    def start(self, config: VMConfig, layout: VMBundleLayout, attach_installer_iso: bool) -> None:
        """Start the VM. attach_installer_iso mirrors the lifecycle effects."""
        if self.is_running:
            raise QemuHostError(f"VM '{config.name}' is already running.")
        self.ensure_bundle_artifacts(config, layout)

        self.qmp_port = _free_loopback_port()
        self.serial_port = _free_loopback_port() if config.serial_console_enabled else 0
        # SSH forward only for a debug VM, and never during the install phase
        # (the guest isn't a running system yet).
        self.ssh_port = (
            _free_loopback_port()
            if config.serial_console_enabled and not attach_installer_iso
            else 0
        )
        args = qemu_command_line.build(
            config,
            layout,
            self._toolchain.uefi_code_path,
            attach_installer_iso,
            self.qmp_port,
            self.serial_port,
            self.ssh_port,
            accel=self._accel,
        )

        proc = subprocess.Popen(
            [self._toolchain.system_binary, *args],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.PIPE,
            text=True,
        )
        self._proc = proc

        stderr_box: list[str] = []

        def watch() -> None:
            try:
                stderr_box.append(proc.stderr.read() if proc.stderr else "")
            except Exception:
                stderr_box.append("")
            code = proc.wait()
            self._proc = None
            self.qmp_port = 0
            self.serial_port = 0
            self.ssh_port = 0
            cb = self.on_guest_stopped
            if cb is not None:
                cb(code, _tail(stderr_box[0] if stderr_box else ""))

        threading.Thread(target=watch, daemon=True).start()

        # Surface an immediate startup failure (bad args, KVM refusal) as a
        # thrown error rather than a phantom "running" VM.
        time.sleep(0.75)
        if proc.poll() is not None:
            self.on_guest_stopped = None
            tail = _tail(stderr_box[0] if stderr_box else "")
            raise QemuHostError(f"QEMU failed to start (exit {proc.returncode}): {tail.strip()}")

    def stop(self, grace_seconds: int = 90) -> None:
        """Ask the guest to power down cleanly (ACPI); after grace_seconds
        without an exit, hard-stop. on_guest_stopped carries the final word."""
        proc = self._proc
        if proc is None or proc.poll() is not None:
            return
        try:
            with QmpClient.connect(self.qmp_port, timeout=5.0) as qmp:
                qmp.system_powerdown()
        except (OSError, QmpError):
            pass  # QMP unreachable — fall through to the hard stop below.
        try:
            proc.wait(timeout=grace_seconds)
            return
        except subprocess.TimeoutExpired:
            pass
        self.force_stop()

    def force_stop(self) -> None:
        proc = self._proc
        if proc is None or proc.poll() is not None:
            return
        try:
            with QmpClient.connect(self.qmp_port, timeout=3.0) as qmp:
                qmp.quit()
            proc.wait(timeout=3)
            return
        except (OSError, QmpError, subprocess.TimeoutExpired):
            pass
        try:
            proc.kill()
            proc.wait(timeout=5)
        except (OSError, subprocess.TimeoutExpired):
            pass
